src/mgr/FunAction.cc

```diff
@@ -50,9 +50,7 @@ Mgr::FunAction::dump(StoreEntry* entry)
     debugs(16, 5, MYNAME);
     Must(entry != nullptr);
     if (UsingSmp())
-        storeAppendPrintf(entry, "by kid%d {\n", KidIdentifier);
+        storeAppendPrintf(entry, "---\nkid: %d\n", KidIdentifier);
     handler(entry);
-    if (atomic() && UsingSmp())
-        storeAppendPrintf(entry, "} by kid%d\n\n", KidIdentifier);
```

Comment:
> Francesco: This solution works by having one yaml document per kid. It is legal ... (see an [easy-read document](https://spacelift.io/blog/yaml))

> easy-read document: Finally, triple dots are used to end a document without starting a new one

The current implementation does not match the syntax documented in that easy-read document because the current implementation does not emit `...` at the end of the output. This is fairly important[^termination] and would have to be fixed if this solution/PR is to be merged. However, this is just a technical bug. It may not be trivial to address in the current code, but it can be addressed. It does _not_ affect the primary format choice.

[^termination]: Without a termination sequence, it is difficult (and, in some contexts, impossible) to know that the entire YAML output has been received in full (i.e. all documents _and_ no truncated document).

Reply:
It turns out that the `...` end of documents marker is optional: https://www.yaml.info/learn/document.html 
It is useful, but IMO not worth the effort at this time (if ever)

Reply:
Sure. The spec is at https://yaml.org/spec/1.2.2/#91-documents
The key sections from it are:
> A document may be preceded by a prefix specifying the character encoding and optional comment lines [..]
> use of two special marker lines [..] one at the start of a document and one at the end. [...]
> At the start of a document, lines beginning with a “%” character are assumed to be directives. The (possibly empty) list of directives is terminated by a directives end marker line. Lines following this marker can safely use “%” as the first character.[...]
> `c-directives-end ::= "---"`
> `c-document-end ::= "..."    # (not followed by non-ws char)`
> [..] The existence of this **optional** document suffix does not necessarily indicate the existence of an actual following document.
> [..]
> A YAML stream consists of zero or more documents. Subsequent documents require some sort of separation marker line. If a document is not terminated by a document end marker line, then the following document must begin with a directives end marker line.


The spec is not very formal in how it is written; it does include BNF, but the way I read it is:
- a stream is made of documents
- each document has an optional prefix (to specify encoding etc), a start-document directive "---" and an optional end-document directive ("..."). 

So for our purposes, to get yaml-compliance all we need to do is replace "by kid <#> {" with "---\nkid: <#>" and "} by kid <#>" with "..."

I've run this through yamllint.com and it was accepted, however the "reformat" feature of yamllint strips the "..." markers, further highlighting that they are optional. I have retested parsing with all libraries I tested above, they all successfully parse the documents

Reply:
Thank you for that research and references! 

I propose that we use the following report syntax (compliant with l-yaml-stream syntax) for our three cases:

```
squid-yaml-stream ::= squid-yaml-bare | squid-yaml-kids | squid-yaml-aggregated
```

... where the three stream variations are defined as follows, one for each of the three primary use cases:

```
squid-yaml-bare ::= l-bare-document
```

squid-yaml-bare is for no aggregation without kid reports (e.g., "mgr:menu"): Each report is represented by a single [l-bare-document](https://yaml.org/spec/1.2.2/#913-bare-documents). This construct keeps these reports unchanged as far as _this_ PR is concerned (other PRs will convert them to YAML).

```
squid-yaml-kids ::= ( l-explicit-document l-document-suffix )*
```

squid-yaml-kids is for kid reports without aggregation (e.g., "mgr:objects"). Each report is an [l-yaml-stream](https://yaml.org/spec/1.2.2/#streams) of l-explicit-document items separated by [with l-document-suffix](https://yaml.org/spec/1.2.2/#rule-l-document-suffix). This format preserves the existing useful begin/end signal for this (streaming with unknown-in-advance content length) use case while changing custom "by kid" prefixes and suffixes to YAML "---" and "..." sequences respectively. That change is being implemented in this PR.

```
squid-yaml-aggregated ::= l-explicit-document
```

squid-yaml-aggregated is for reports aggregated across kids (e.g., "mgr:info"). Each report is an [l-explicit-document](https://yaml.org/spec/1.2.2/#914-explicit-documents). Note that this construct does not end with with [l-document-suffix](https://yaml.org/spec/1.2.2/#rule-l-document-suffix). This construct requires changing affected reports (as they get converted to YAML) to start with YAML "---". No changes for this PR though.

Reply:
Thanks for taking the time to write down this formal proposal; let's talk about it later on today.

In my understanding there is only one change proposed to what is currently being done. In lay-words, what you are suggesting to do is:

- menu has no document markers, as currently done
- non-aggregated reports have each sub-report start with ---, end with ... , as currently done by the ones I've converted
- aggregated reports have a --- document-start marker, but no ... document-end marker . Implementing this is for a future PR

Is this correct, or am I missing anything?

Reply:
> let's talk about it later on today.

I am looking forward to that conversation.

> Is this correct, or am I missing anything?

Your "lay-words" text (uses a mgr:menu _example_ instead of the corresponding concept and misinterprets current PR code state a bit but) does not contradict any major aspects of my proposal AFAICT. It would be foolish of me to confirm or deny that the "lay-words" text (is "correct" in a sense that it) should be implemented: That "lay-words" text may result in missed or misinterpreted implementation _details_. If accepted, I recommend going by the (adjusted as necessary) proposal instead.